common/persistence/sql/sqlplugin/sqlite/db_pool.go

```diff
@@ -0,0 +1,57 @@
+package sqlite
+
+import (
+	"sync"
+
+	"github.com/jmoiron/sqlx"
+)
+
+// SQLite database takes exclusive access to the database file during write operations.
+// If another process attempts to perform a write operation, it receives a "database is locked" error.
+// To ensure only one database connection is created per database file within the running process,
+// we use dbPool to track and reuse database connections.
+// When a new database connection is requested, it increments dbPoolCounter and returns the existing sql.DB
+// from dbPool. When a connection is requested to be closed, it decrements dbPoolCounter.
+// Once the counter reaches 0, the database connection is closed as there are no more references to it.
+// Reference: https://github.com/mattn/go-sqlite3/issues/274#issuecomment-232942571
+var (
+	dbPool        = make(map[string]*sqlx.DB)
```

Comment:
Since we're not going to be opening a db, using it for a bit, then closing it and opening a _different_ one: should we just open a connection at process start (per database), and never close it?

Reply:
I'd love to do it; however, there are a few challenges that make it quite complex.

To provide one dependency, we have to refactor the server initialization logic. Each service (frontend/history/...) is a standalone `fx.App`, and their initialization is independent. Technically, merging them is possible; however, it may require some extra work to support this particular use case.

During initialization, a new connection to the database is created not only for the Datastore but also for version compatibility checks, which causes concurrent access to the SQLite database. This logic must also be migrated to fx to prevent it.

So, the solution above helps us avoid the refactoring and guarantees that only one database connection is opened per SQLite database file.